Refactoring api: changed `login`, `refresh` endpoints

Hrushon opened the pull request:
Changed `/login` & `/refresh` endpoints api's.
Now they return dictionary includes access token and info about user-admin.
In addition, they return refresh token inside cookies.
Changed schemas for represent information about the administrator.
Added DTO model for tokens and administrator instance.

https://www.notion.so/Refactoring-POST-administrators-login-administrators-refresh-606b27ee83dd4ca387ca53ecbcb8a429

Commit:
refactor(api): changed `login`, 'refresh` endpnts

Changed `/login` & `/refresh` endpoints api's. Now they return
dictionary includes access token and info about user-admin.
In addition, they return refresh token inside cookies.

ADesBiysk commented on src/api/response_models/administrator.py:
```diff
@@ -25,4 +25,4 @@ class TokenResponse(BaseModel):
 
     access_token: str
-    refresh_token: str
+    refresh_token: str | None
```
Чтобы не вводить в заблуждение, в теле ответа этого поля не должно быть, даже пустого.

ADesBiysk replied:
И добавь в эту модель Администратора, тогда не надо будет описывать странную response_model как словарь.
Если это что-то ломает, то лучше добавить новую.

ADesBiysk commented on src/api/routers/administrator.py:
```diff
@@ -26,31 +27,47 @@ class AdministratorCBV:
         - **password**: пароль
         """
-        return await self.authentication_service.login(auth_data)
+        tokens = await self.authentication_service.login(auth_data)
```
Получать токены нужно через дополнительную DTO модель.

ADesBiysk commented on src/api/routers/administrator.py:
```diff
@@ -26,31 +27,47 @@ class AdministratorCBV:
         """
-        return await self.authentication_service.login(auth_data)
+        tokens = await self.authentication_service.login(auth_data)
+        administrator = await self.authentication_service._AuthenticationService__authenticate_administrator(auth_data)
```
Очень плохое решение:
1. Получаем доступ к приватному методу класса обходными путями.
2. Вызываем один и тот же метод 2 раза подряд. 

Здесь тоже поможет DTO модель. Просто в неё нужно кроме токенов добавить ещё и модель Administrator.

Hrushon replied:
Исправил
Берут сомнения, что сделал не то, что требовалось.

ADesBiysk commented on src/api/routers/administrator.py:
```diff
@@ -26,31 +27,47 @@ class AdministratorCBV:
-        """Обновление access и refresh токенов при помощи refresh токена."""
-        return await self.authentication_service.refresh(request_data.refresh_token)
+    async def refresh(
+        self, response: Response, request_data: RefreshToken
```
Дополнил задачу. Токен обновления получает также через куки. В теле её присылать не будут.

ADesBiysk commented on src/api/routers/administrator.py:
```diff
@@ -26,31 +27,47 @@ class AdministratorCBV:
+    async def refresh(
+        self, response: Response, request_data: RefreshToken
+    ) -> dict[str, Union[TokenResponse, AdministratorResponse]]:
+        """Обновление access и refresh токенов при помощи refresh токена.
```
Здесь всё аналогично методу login.

Commit:
refactor(api): changed `login`, `refresh` endpnts

Changed `/login` & `/refresh` endpoints api's.
Now they return dictionary includes access token and info about user-admin.
In addition, they return refresh token inside cookies.
Changed schemas for represent information about the administrator.

ADesBiysk commented on src/core/services/authentication_service.py:
```diff
@@ -82,10 +81,9 @@ async def get_current_active_administrator(self, token: str) -> Administrator:
 
-    async def refresh(self, token: str) -> TokenResponse:
-        """Метод получения новой пары refresh- и access- токенов."""
+    async def refresh(self, token: str) -> TwoTokensAdministratorResponse:
```
Authentification_service возвращает DTO модель. Например, такую:
```
@dataclass
class AdministratorAndTokens:
    access_token: str
    refresh_token: str
    administrator: Administrator
```
Располагаются они src/core/db/DTO_models.py

ADesBiysk commented on src/core/services/authentication_service.py:
```diff
@@ -58,15 +58,14 @@ async def __authenticate_administrator(self, auth_data: AdministratorAuthenticat
-            access_token=self.__create_jwt_token(administrator.email, ACCESS_TOKEN_EXPIRE_MINUTES),
-            refresh_token=self.__create_jwt_token(administrator.email, REFRESH_TOKEN_EXPIRE_MINUTES),
-        )
+        administrator.access_token = self.__create_jwt_token(administrator.email, ACCESS_TOKEN_EXPIRE_MINUTES)
+        administrator.refresh_token = self.__create_jwt_token(administrator.email, REFRESH_TOKEN_EXPIRE_MINUTES)
+        return administrator
```
```suggestion
        await self.__administrator_repository.update(administrator.id, administrator)
        return AdministratorAndTokens(
            access_token=self.__create_jwt_token(administrator.email, ACCESS_TOKEN_EXPIRE_MINUTES),
            refresh_token=self.__create_jwt_token(administrator.email, REFRESH_TOKEN_EXPIRE_MINUTES),
            administrator=administrator,
        )
```

ADesBiysk commented on src/api/routers/administrator.py:
```diff
@@ -26,31 +29,45 @@ class AdministratorCBV:
         - **password**: пароль
         """
-        return await self.authentication_service.login(auth_data)
+        data = await self.authentication_service.login(auth_data)
+        response.set_cookie(key="refresh_token", value=data.refresh_token, httponly=True, samesite="strict")
+        return data
```
```suggestion
        admin_and_token = await self.authentication_service.login(auth_data)
        response.set_cookie(key="refresh_token", value=admin_and_token.refresh_token, httponly=True, samesite="strict")
        result = AdministratorAndAccessTokenResponse.from_orm(admin_and_token.administrator)
        result.access_token = admin_and_token.access_token
        return result
```

ADesBiysk commented on src/api/routers/administrator.py:
```diff
@@ -26,31 +29,45 @@ class AdministratorCBV:
-        """Аутентифицировать администратора по email и паролю.
+    async def login(
+        self, response: Response, auth_data: AdministratorAuthenticateRequest
+    ) -> TokenAdministratorResponse:
```
```suggestion
    ) -> AdministratorAndAccessTokenResponse:

```

```
class AdministratorAndAccessTokenResponse(AdministratorResponse):
    """Схема для отображения access-токена и информации об администраторе."""

    access_token: str | None
```

ADesBiysk reviewed (commented):
Немного кривовато, но примерно так.

Commit:
refactor(api): changed `login`, `refresh` endpnts

Changed `/login` & `/refresh` endpoints api's.
Now they return dictionary includes access token and info about
user-admin.
In addition, they return refresh token inside cookies.
Changed schemas for represent information about the administrator.
Added DTO model for tokens and administrator instance.

Hrushon committed:
Merge branch 'develop' into refactoring/post_administrators
